### ISSUE TITLE
Joining through a SteamID causes some desync.

### ISSUE THREAD

**Issue by Totoqoe** [winstxnhdw/lc-hax#369]
### What happened?

Using the join via ID feature, all the furniture is desynced and are placed on the default positions regardless of the placements the host has.  It also lists all player names as unknown and the invite menu seems to break.  Even when using the Better_Lobbies mod, you're unable to get the Lobby ID (normally starts with a 1 such as 109775241586324425).

### Current Commit Hash

f195ff19193bb67ca0d1dd174087f0578a14a5d5

### Injector

- [X] SharpMonoInjectorCore
- [ ] Others

### Acknowledgement

- [X] I have confirmed that my anti-virus is not blocking any of the relevant programs
- [X] I have done my due diligence to check for similar issues

**Comment by Totoqoe** [comment]
It also causes intense frame drops when you join with antikick, to the point of dipping to ~10fps

**Comment by winstxnhdw** [comment]
I’ll look into this but the issue with the FPS is probably due to exceptions. I can’t fix it but I can tell you how to alleviate it. If your BepInEx logs are printing a bunch of exceptions, try restarting and injecting only when you are on the start menu screen. You should see that the exceptions disappear.

**Comment by Totoqoe** [comment]
> I’ll look into this but the issue with the FPS is probably due to exceptions. I can’t fix it but I can tell you how to alleviate it. If your BepInEx logs are printing a bunch of exceptions, try restarting and injecting only when you are on the start menu screen. You should see the exceptions disappear.

The start menu as in where it says Online and LAN?

**Comment by winstxnhdw** [comment]
The menu after that.

**Comment by Totoqoe** [comment]
Ah, that's where I normally inject it (assuming you mean where it allows you to host or join lobbies)

**Comment by winstxnhdw** [comment]
Yup.

**Comment by Totoqoe** [comment]
To add to this, this seems to happen whenever you join through the connect feature, even if you were in the lobby before (such as being reconnected from anti kick).  All the furniture still gets desynced, the usernames do get cached but any new players that join will receive the "unknown" tag, and the invite menu and lobby id gets broken.

**Comment by winstxnhdw** [comment]
> the usernames do get cached but any new players that join will receive the "unknown" tag, and the invite menu and lobby id gets broken.

We know this and it has been discussed countless times. There’s no way to fix it and we don’t have plans to investigate further.

**Comment by Turkeii** [comment]
You mean you cant fix it, considering you cant make a UI its not surprising.

**Comment by Totoqoe** [comment]
> You mean you cant fix it, considering you cant make a UI its not surprising.

This comment unnecessary, I believe this is just a limitation with how the lobbies work and the side effects of essentially bypassing the lobby ID to join directly to a steam ID.

**Comment by Turkeii** [comment]
> > You mean you cant fix it, considering you cant make a UI its not surprising.
> 
> This comment is unnecessary, I believe this is just a limitation with how the lobbies work and the side effects of essentially bypassing the lobby ID to join directly to a steam ID.

By any chance when you join a lobby and it sets everyone's name to unknown you have Anti Kick enabled?

**Comment by Totoqoe** [comment]
> > > You mean you cant fix it, considering you cant make a UI its not surprising.
> > 
> > 
> > This comment is unnecessary, I believe this is just a limitation with how the lobbies work and the side effects of essentially bypassing the lobby ID to join directly to a steam ID.
> 
> By any chance when you join a lobby and it sets everyone's name to unknown you have Anti Kick enabled?

yes it does that whenever the rejoin triggers or when you join directly via a steam ID.

**Comment by Turkeii** [comment]
You keep saying steamId do you mean a lobbyId?

**Comment by Totoqoe** [comment]
The Lobby ID is a seperate thing from a steam ID.  I believe the Lobby ID is how the game properly syncs players and objects before it routes you to the host for the connection.  For example, the bepinex logs shows you three parameters when you join through an invite link. 
AA *game ID*
BB *Lobby ID
CC *Steam ID*
I believe these are the paremeters used in the invite link. The CC is the host's steam ID.
Now if you were to join using the rejoin feature patch or the join through steam ID patch, it will not input those parameters and will simply skip to the steam ID as shown in the logs.
"CC *steamID*"

**Comment by Totoqoe** [comment]
Small correction, it seems the AA and BB parameter are both the lobby ID as shown here.
![image](https://github.com/winstxnhdw/lc-hax/assets/158114082/67d87a45-0cff-41ef-8453-b9518b18c680)
And heres what it looks like when connecting via the steam ID or when triggering a rejoin..
![image](https://github.com/winstxnhdw/lc-hax/assets/158114082/568689ed-0418-4fef-b6a9-c06f77ed9ae6)

**Comment by Turkeii** [comment]
🤦🏻 You shouldn't try to act like you know what you're talking about, its the function he's calling that the issue is coming from should. He should instead use JoinLobby that uses Lobby instead as it calls the right functions. CC is the Owners Steam Id AA and BB are both the lobby id. // I just see you updated your comment.

**Comment by Totoqoe** [comment]
> 🤦🏻 You shouldn't try to act like you know what you're talking about, its the function he's calling that the issue is coming from should. He should instead use JoinLobby that uses Lobby instead as it calls the right functions. CC is the Owners Steam Id AA and BB are both the lobby id. // I just see you updated your comment.

My bad, I'm just trying my best to understand the game with little to no knowledge of game code lol

**Comment by Turkeii** [comment]
I get not fully understanding it without looking at the code so ill explain the issue. The function he calls is StartClient which just takes the LobbyId Does not call the right things to sync as when you physically click a lobby and join it calls a separate function that calls StartClient after initializing the lobby values hence why you cant get the lobbyId when joining back.

**Comment by Turkeii** [comment]
![image](https://github.com/winstxnhdw/lc-hax/assets/81539722/54941aa2-8f76-4435-a563-832668fb9079)

**Comment by Turkeii** [comment]
You can make a Lobby by using a SteamId anyways so it should be no hassle changing 1 line of code.

**Comment by Totoqoe** [comment]
Should I reopen the issue then? Would this also make it possible to grab the lobby ID when joining only through a steam ID? (for example, to join a private lobby)

**Comment by Turkeii** [comment]
I don't care what you make the issue I'm just offering a solution and explanation to both you and Winston.
"GameNetworkManager.Instance.JoinLobby(new Lobby(lobbyId), lobbyId);" This is what he'd have to call for it to sync correctly.

**Comment by Totoqoe** [comment]
@winstxnhdw sorry to bother ya but would this be a possible fix for this issue?

**Comment by winstxnhdw** [comment]
I tried that method while trying to fix #332. Doesn’t seem to work. I even tried finding the exact lobby component and using their public `Join` method. Still no go. 

@Turkeii it’s best if you test it yourself before reading a bunch and assuming you know what works.

**Comment by winstxnhdw** [comment]
> You mean you cant fix it

@Turkeii the unknown nametag happens regardless of auto-rejoin and happens because we don’t call the OnConnectedPlayer server RPC. I’d really appreciate it if you’d do a tiny bit of research before bumping an old issue and wasting everyone’s time like a skid. 

If you think GUIs are a better design than stateless commands, I got something else to tell you 😂

**Comment by Turkeii** [comment]
> @Turkeii it’s best if you test it yourself before assuming what works.

I find this oxymoronic and i ask that you please let your ego take a hit on this one and accept that you are wrong.
If you are so delusional that i need to attach a video and showcase it working then perhaps i will.

**Comment by winstxnhdw** [comment]
I am more than happy to see a video. It’ll save all of us time :)

**Comment by Turkeii** [comment]
> I am more than happy to see a video. It’ll save all of us time :)
> 
> By the way, this is GitHub and not UC. Collaborate professionally and reduce the hormone dosage. There is so much unprovoked rage in your words.

Very hypocritical of you, I can see all the messages you edited lmfao.

**Comment by winstxnhdw** [comment]
Exactly. That's why.. they're edited? To be nice.

**Comment by Turkeii** [comment]
What happens when i prove you wrong?

**Comment by winstxnhdw** [comment]
Well, you get to prove me wrong, and make thousands of people happier because now they'll have a better auto-rejoin.

**Comment by Turkeii** [comment]
> I am more than happy to see a video. It’ll save all of us time :)

https://youtu.be/wzUDY2wDP_0

**Comment by winstxnhdw** [comment]
This is with anti-kick?

**Comment by Turkeii** [comment]
> This is with anti-kick?

No why would you think this?

**Comment by winstxnhdw** [comment]
Because only with anti-kick does this issue occur and that's what @Totoqoe has been saying talking about?

**Comment by winstxnhdw** [comment]
Of course, there wouldn't be any issues without anti-kick.

**Comment by Turkeii** [comment]
> I tried that method while trying to fix #332. Doesn’t seem to work. I even tried finding the exact lobby component and using their public `Join` method. Still no go. Anyways, I have kept that issue open because I still haven’t exhausted all the options.
> 
> And @Totoqoe, there’s no need to apologise to them. Your understanding is probably correct. My guess is that `LobbyId` is not the same as `SteamId`, that’s why those methods failed.
> 
> @Turkeii it’s best if you test it yourself before assuming what works.

I'm referring to this.

**Comment by Turkeii** [comment]
And the de sync you are having is also related to the function your calling.

**Comment by winstxnhdw** [comment]
So the furniture syncing issue? There's no furniture in your video though? Just to be sure, can you record one with furniture and place them at random locations. If your method works, we should see the furniture back in the default position.

**Comment by Turkeii** [comment]
> So the furniture syncing issue? There's no furniture in your video though? Just to be sure, can you record one with furniture and place them at random locations. If your method works, we should see the furniture back in the default position.

That's how the lobby was that i joined a random lobby.

**Comment by winstxnhdw** [comment]
Nvm, I can just test it myself.

**Comment by Turkeii** [comment]
> @Turkeii it’s best if you test it yourself before assuming what works.

Iconic comment.

**Comment by winstxnhdw** [comment]
I've tested it before, I am just doing it again.

**Comment by Turkeii** [comment]
![image](https://github.com/winstxnhdw/lc-hax/assets/81539722/98ccfa66-e9e5-4cac-9502-694ddc275615)
doubt this would help since you don't have a module system let alone a real setting system either but this is the code.

**Comment by winstxnhdw** [comment]
Yeah, as expected. Same thing.

![image](https://github.com/winstxnhdw/lc-hax/assets/56998716/30dabdb0-f914-4771-ae7e-ddf80df14768)

**Comment by Turkeii** [comment]
> Yeah, as expected. Same thing.
> 
> ![image](https://private-user-images.githubusercontent.com/56998716/305649127-30dabdb0-f914-4771-ae7e-ddf80df14768.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgxNzk4OTksIm5iZiI6MTcwODE3OTU5OSwicGF0aCI6Ii81Njk5ODcxNi8zMDU2NDkxMjctMzBkYWJkYjAtZjkxNC00NzcxLWFlN2UtZGRmODBkZjE0NzY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE3VDE0MTk1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlNThjOGVhMWFlNTQyMTQyMThjYzYyYzBhZTU2YzU1YTAzNTg4ODg5OTRkZTQ5ZGM4Mjk5YmVlYTEwNGM5NTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.fylmAa3TZKfe9OfDlGbFtzCCUXjp43HKnAkIdlkRchU)

Seems like a skill issue.

**Comment by winstxnhdw** [comment]
Doubt it.

**Comment by Turkeii** [comment]
> Doubt it.

Obviously you watched my video and saw it work.

**Comment by winstxnhdw** [comment]
Well well well, looks like I am the one that's right in the end. The reason why I had that issue was because I was using the Steam ID as the Lobby ID. When I copied the Lobby ID instead, it worked.

**Comment by winstxnhdw** [comment]
> doubt this would help since you don't have a module system let alone a real setting system either but this is the code.

Out of curiosity, why do you think have a setting system or a GUI makes your codebase that much better? I don't get what's so "superior" about holding state.

Is this your codebase?

![image](https://github.com/winstxnhdw/lc-hax/assets/56998716/dbea675a-11d1-47b2-9951-6d5d0b3cca07)

I've worked with some Vietnamese engineers before, they also loved writing code like that but I figured it was because they were paid too little to care.

**Comment by winstxnhdw** [comment]
@Totoqoe Anyways, I am not sure how we can move on from here. My intention to use Steam ID was that Steam ID's are publicly available unlike Lobby IDs. If I move over to using Lobby IDs permanently, we won't be able to join via Steam IDs anymore.

Maybe I can fallback to using Lobby ID for auto-rejoin but for the clipboard, I'll accept both Steam and Lobby IDs?

**Comment by OldGreggo5081** [comment]
> Well well well, looks like I am the one that's right in the end. The reason why I had that issue was because I was using the Steam ID as the Lobby ID. When I copied the Lobby ID instead, it worked.

How were you right in the end, you denounced someone for doing it correctly lol?

**Comment by winstxnhdw** [comment]
> You can make a Lobby by using a SteamId anyways so it should be no hassle changing 1 line of code.

Because he mentioned this.

**Comment by OldGreggo5081** [comment]
but you were still wrong in the beginning?

**Comment by winstxnhdw** [comment]
It's not wrong? As mentioned,

> My intention to use Steam ID was that Steam ID's are publicly available unlike Lobby IDs. If I move over to using Lobby IDs permanently, we won't be able to join via Steam IDs anymore.

You can't search up lobby IDs and join them. By using Lobby IDs, you are removing a feature to search up a person's Steam ID and joining them. That has always been the intention.

**Comment by OldGreggo5081** [comment]
Im not looking to argue im just confused to why you saw the need to not act professionally and try reclaim the high ground after being proved wrong

**Comment by OldGreggo5081** [comment]
everyone is wrong at some point, and its ok to admit their mistakes.

**Comment by winstxnhdw** [comment]
Hey man, if you want to be right so much, you can have it. But it's nice to be fair.

**Comment by Turkeii** [comment]
> I've worked with some Vietnamese engineers before, they also loved writing code like that but I figured it was because they were paid too little to care.

Ironic I'm not even Vietnamese yet you chose to say an insult after proved wrong.
And to respond to your picture of someones shitty UI made with unity's GUI that is nothing like mine, thank you for reminding me how shitty other peoples UI is.

**Comment by winstxnhdw** [comment]
Sorry, I didn't mean it as an insult.

**Comment by Turkeii** [comment]
Is it easier to type words than it is to click your mouse to toggle modules/settings?

**Comment by winstxnhdw** [comment]
It's preference isn't it. It's why engineers use the CLI more than they use a GUI. There's a level of control that's harder to attain with something that abstracts so much away from the user. 

Of course, if your users are all below 18, a GUI is much preferred.

**Comment by Turkeii** [comment]
Age doesn't matter when it comes to simplicity within using a menu.
If you'd like we can take this discussion on to discord instead of an issue topic.

**Comment by winstxnhdw** [comment]
Again, it's just preference. My mod allows you to easily repeat commands with the up arrow, similar to the terminal.

Anyways, take the poison command for example.

```console
/poison <player> <damage> <duration> <delay=1>
```

How would this be done in a GUI? A button with 4 sliders?

How about `grab`?

```console
/grab <item?>
```

A button with a field? Which also means they'd have to end up typing anyways. In the end, your GUI ends up with pages of cluttered nonsense that the user has to search through. Not a very pleasant experience isn't it? Reminds me of those Minecraft cheats that cover your entire screen. Half of them are usually useless and would never enable.

I don't think there anymore we can add to this discussion. If you strongly think otherwise, you can leave your Discord and we can continue our conversation.

**Comment by Turkeii** [comment]
There's a use for both, how do you suppose you are gonna do this in commands?
![image](https://github.com/winstxnhdw/lc-hax/assets/81539722/f6d67972-b89c-4389-8467-f233df3d8c1a)
With the user inputting their colors in by hand?

**Comment by Turkeii** [comment]
Changing key binds could be as simple as left clicking the (V) and pressing the new key bind. Vs how you'd do it inside of a command.

**Comment by winstxnhdw** [comment]
> There's a use for both, how do you suppose you are gonna do this in commands?

This is for..? To make my non-existent GUI pretty? Only valid reason why someone would need to do this is if they wanted to customise their ESP colours. Which is fair, but it's not a priority of this project. Instead, we want to reduce the mental overhead of customisability. I'd understand if the mod was meant to be some appearance modifier mod, but it isn't.

> Changing key binds could be as simple as left clicking the (V) and pressing the new key bind. Vs how you'd do it inside of a command.

Why reinvent the wheel? If I wanted to allow users to change their keys, I'd reuse and extend Lethal Company's existing key changer.

**Comment by Turkeii** [comment]
> In the end, your GUI ends up with pages of cluttered nonsense that the user has to search through. Not a very pleasant experience isn't it? Reminds me of those Minecraft cheats that cover your entire screen. Half of them are usually useless and would never enable.
Would never enable?
I think your reaching with "pages of cluttered nonsense" pretty easy to find what you're looking for in my Gui's cant same the same for other Lethal menus.

**Comment by winstxnhdw** [comment]
![image](https://github.com/winstxnhdw/lc-hax/assets/56998716/6855193c-367c-4916-8531-5eae6100db95)

Your tabs literally have subtabs...

**Comment by Turkeii** [comment]
> ![image](https://private-user-images.githubusercontent.com/56998716/305653302-6855193c-367c-4916-8531-5eae6100db95.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgxODQzNjUsIm5iZiI6MTcwODE4NDA2NSwicGF0aCI6Ii81Njk5ODcxNi8zMDU2NTMzMDItNjg1NTE5M2MtMzY3Yy00OTE2LTg1MzEtNWVhZTYxMDBkYjk1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE3VDE1MzQyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlN2IzZjU3MjgxODA2NzIyZDZmNTQ1MzM5YWFmOWFkZTA1YzhmNjBjYTllZWI4YmVmOTkxYjJhNWQ5ZWQ4YjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DQvIehPjsy14T56Lur1lrddid7hBWyPlAiJRANppNN4)
> 
> Your tabs literally have subtabs...

What does sub tabs prove?

**Comment by winstxnhdw** [comment]
"Oh yes, let me enable god mode. Hold on.. is it in players? Oh it's in self.. In the damage subtab, wait no, it's in the status subtab?"

**Comment by Turkeii** [comment]
Wtf lmfao, Godmode is a module inside of the category self and to enable it all you have to do is click it once.

**Comment by Turkeii** [comment]
Players is for the players in the lobby.

**Comment by winstxnhdw** [comment]
Yeah, it's just a funny example. What I really mean is that you know your cheats well so of course you'd know where everything is.

**Comment by Turkeii** [comment]
> Yeah, it's just a funny example. What I really mean is that you know your cheats well so of course you'd know where everything is.

And you'd know each command right off the bat of using your menu?

**Comment by winstxnhdw** [comment]
`ctrl + F` works wonders.

**Comment by Turkeii** [comment]
Pressing left click works wonders.

**Comment by winstxnhdw** [comment]
"Bro, he's finally wearing a jetpack, quickly nuke it."

"Ok, hold on, I need to find the explode option under the players tab, under his player subtab, under the explosion category."

**Comment by Turkeii** [comment]
Do you know what a key bind is?

**Comment by winstxnhdw** [comment]
Let's see how many keybinds you can add as you scale.

**Comment by Turkeii** [comment]
> Btw, we also use keybinds.
Congratulations?
How many keybinds can you add?

**Comment by winstxnhdw** [comment]
Not as many, because most of them can just be commands.

**Comment by Turkeii** [comment]
> Not as many, because most of them can just be commands.

And most of mine can be buttons

**Comment by Turkeii** [comment]
How do you do commands in the main menu?

**Comment by winstxnhdw** [comment]
You don't because you don't have to. Beautiful, am I right? It implicitly deactivates. Meanwhile you:

if not in game: return
if object is null: return

**Comment by Turkeii** [comment]
> You don't because you don't have to. Beautiful, am I right? It implicitly deactivates. Meanwhile you:
> 
> if not in game: return if object is null: return

You're speaking on things you have no understanding of and its funny to see you try to degrade my code and compare yours to mine yet they are 2 entirely different things.

**Comment by winstxnhdw** [comment]
Make your code public then? Or we can just agree to disagree.

**Comment by Turkeii** [comment]
> Make your code public then? Or we can just agree to disagree.

I'm alright with my code being private as you people would only take from it.

**Comment by winstxnhdw** [comment]
LMFAO, you mean the other way around? Because aside from `revive` and infinite grab distance, everything has been written from scratch.

**Comment by Turkeii** [comment]
> LMFAO, you mean the other way around? Because aside from `revive` and infinite grab distance, everything has been written from scratch. If you read from any of the other projects, you've probably indirectly copied me too.

What could i possibly take from you that i already haven't found? 
The game is open source its bound to have 2 people having the same code but if you ever saw what i have you'd take it.

**Comment by winstxnhdw** [comment]
> What could i possibly take from you that i already haven't found?

Exactly, because my project is open-source and I don't start crying because someone use my code.

**Comment by Turkeii** [comment]
> > What could i possibly take from you that i already haven't found?
> 
> Exactly, because my project is open-source, you've already found it. And I don't start crying because someone uses/sells my code. Pretty sure you never had `destroy` either until you looked here.

Congratulations you found a function inside of a class that can be seen by anyone who looks.

**Comment by OldGreggo5081** [comment]
so much for being professional lol

**Comment by winstxnhdw** [comment]
> Congratulations you found a function inside of a class that can be seen by anyone who looks.

Yet no one could get it to work before me. It's how you use it more than finding it.

**Comment by winstxnhdw** [comment]
> so much for being professional lol

What's not professional here?

**Comment by Turkeii** [comment]
> > Congratulations you found a function inside of a class that can be seen by anyone who looks.
> 
> Yet no one could get it to work before me. It's how you use it more than finding it.

So hard to call a function 😢

**Comment by OldGreggo5081** [comment]
> > so much for being professional lol
> 
> What's not professional here?

you, you are actively responding to harsh criticism in a negative way, any person with common sense would've closed the issue or... blocked them. Thank you for understanding, also regarding the destroy thing, are you referring to the function inside of the PlayerControllerB class that destroys the item in hand?

**Comment by winstxnhdw** [comment]
Your reaction to harsh criticism is to block someone? What a way to live in an echo chamber..

**Comment by OldGreggo5081** [comment]
well, he seems to be getting on your nerves, and considering you were "right" in the end, why do you persist and talk to them?

**Comment by OldGreggo5081** [comment]
its called common sense, if someone is bullying you, then block them.

**Comment by winstxnhdw** [comment]
Because that’s just not the right thing to do? In the end, I am trying to understand what he has to convey. I mean if he was only being an asshole, sure. But he’s kind of reasonable now, although it has gone worse now.

**Comment by OldGreggo5081** [comment]
idk, if it were me and someone wasnt being "nice and fair" and resorted to name calling after giving criticism then i would've blocked them 🤷

**Comment by OldGreggo5081** [comment]
plus considering the topic of this issue has deviated so far from its original topic, I would've closed it by now 🤷‍♂️ just a thought
